kgforge/core/commons/sparql_query_builder.py

```diff
@@ -345,20 +367,120 @@ def _replace_in_sparql(
+        elif not isinstance(statements, list):
+            raise AttributeError(
+                f"Expected statements argument to be of type  List[str] when union is True. Each item will go in a UNION clause: {statements}"
+            )
+        else:
+            all_statements = "}} UNION {{".join(statements)
+            all_statements = "{{" + all_statements + "}}"
+
+        where_clauses = (
+            f"{{ Graph ?g {{{all_statements}}}}}"
+            if search_in_graph
+            else f"{{{all_statements}}}"
+        )
+
+        join_vars_ = " ".join(vars_)
+        select_vars = f"DISTINCT {join_vars_}" if distinct else f"{join_vars_}"
+        return f"SELECT {select_vars} WHERE {where_clauses}"
+
 
 def _box_value_as_full_iri(value):
     return f"<{value}>" if is_valid_url(value) else value
+
+
+def build_shacl_query(
+    statements: List[str] = None,
+    defining_property_uri: str = None,
+    deprecated_property_uri: str = None,
+    deprecated: bool = False,
+    deprecated_optional: bool = False,
+    search_in_graph: bool = False,
+    context: Context = None,
+) -> str:
+    deprecated_statement = (
+        f"?resource_id <{deprecated_property_uri}> '{str(deprecated).lower()}'^^xsd:boolean"
+        if deprecated_property_uri
+        else ""
+    )
+    defining_statement = (
+        f"?resource_id <{defining_property_uri}> ?shape"
+        if defining_property_uri
+        else ""
+    )
+    if deprecated_optional and deprecated_property_uri:
+        deprecated_statement = " OPTIONAL {" + deprecated_statement + "}"
+    shape_target_statement = "?shape sh:targetClass ?type"
+    extra_statements = [defining_statement, deprecated_statement]
+    if statements:
+        extra_statements.extend(statements)
+    shape_target_statement_str = ".".join([shape_target_statement] + extra_statements)
+    shape_node_statements_str = (
+        "SELECT (?shape as ?type) ?shape ?resource_id WHERE { ?shape a sh:NodeShape ."
+        + "?shape a rdfs:Class . {}".format(".".join(extra_statements) + "}")
+    )
+    shape_node_not_targeting_statements_str = (
+        "SELECT ?shape ?resource_id WHERE { ?shape a sh:NodeShape ."
+        + "FILTER NOT EXISTS {?shape sh:targetClass []} FILTER NOT EXISTS {?shape a rdfs:Class}"
+        + " . {}".format(".".join(extra_statements) + "}")
+    )
+    all_statements = [
+        shape_target_statement_str,
+        shape_node_statements_str,
+        shape_node_not_targeting_statements_str,
+    ]
+    vars_ = ["?type", "?shape", "?resource_id"]
+    if search_in_graph:
+        vars_.append("?g")
+    shacl_query = SPARQLQueryBuilder.create_select_query(
+        vars_=vars_,
+        statements=all_statements,
+        distinct=True,
+        search_in_graph=search_in_graph,
+        union=True,
+    )
+    if not context:
+        return shacl_query
+    else:
+        return SPARQLQueryBuilder.rewrite_sparql(
+            shacl_query,
+            context_as_dict=context.document,
```

Comment:
You should look into the way the read only store uses rewrite_sparql. 

It seems there needs to be a specific shape for context_as_dict to have 
(using _context_to_dict)

Reply:
for the query built by build_shacl_query, context.document is enough as only adding sh and rdfs prefixes are needed. Now the jsonld context processing logic in [here](https://github.com/BlueBrain/nexus-forge/blob/master/kgforge/core/archetypes/read_only_store.py#L58) should be brought back in SPARQLQueryBuilder.rewrite_sparql which should accept a Context